### ISSUE TITLE
compile error on Debian GNU/Linux 10 (buster) Open3D v.0.11.0

### ISSUE THREAD

**Issue by nanotuxi** [intel-isl/Open3D#2487]
**Describe the bug**
Trying to compile Open3d from git (master) v0.11.0 in a python venv throws an error
related to 
`Open3D/cpp/open3d/geometry/Line3D.h:117:9: error: ‘transform’ was not declared in this scope`

**To Reproduce**
1. Clone a fresh version of Open3d 0.11.0
`git clone --recursive https://github.com/intel-isl/Open3D`
2. cd Open3D && mkdir build && cd build
3. configure with `ccmake ..`
4. run `cmake ..`
5. run `make -j6`
This throws 
```
Scanning dependencies of target visualization
[ 55%] Building CXX object cpp/open3d/geometry/CMakeFiles/geometry.dir/Line3D.cpp.o
[ 55%] Building CXX object cpp/open3d/geometry/CMakeFiles/geometry.dir/MeshBase.cpp.o
[ 55%] Building CXX object cpp/open3d/geometry/CMakeFiles/geometry.dir/IntersectionTest.cpp.o
[ 55%] Building CXX object cpp/open3d/geometry/CMakeFiles/geometry.dir/TriangleMesh.cpp.o
[ 56%] Building CXX object cpp/open3d/geometry/CMakeFiles/geometry.dir/SurfaceReconstructionBallPivoting.cpp.o
[ 57%] Building CXX object cpp/open3d/geometry/CMakeFiles/geometry.dir/VoxelGridFactory.cpp.o
[ 57%] Building CXX object cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/RGBDImageShader.cpp.o
In file included from ~/Open3D/cpp/open3d/geometry/IntersectionTest.h:32,
                 from ~/Open3D/cpp/open3d/geometry/SurfaceReconstructionBallPivoting.cpp:31:
~/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Line3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
~/Open3D/cpp/open3d/geometry/Line3D.h:117:9: error: ‘transform’ was not declared in this scope
         transform(t);
         ^~~~~~~~~
//reduce output error size to shorten reading manually
[... cut original error output manually...]

In file included from ~/Open3D/cpp/open3d/Open3D.h:40,
                 from ~/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.cpp:29:
~/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Line3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
~/Open3D/cpp/open3d/geometry/Line3D.h:117:9: error: ‘transform’ was not declared in this scope
         transform(t);
         ^~~~~~~~~
~/Open3D/cpp/open3d/geometry/Line3D.h:117:9: note: suggested alternative:
In file included from /usr/include/c++/8/algorithm:62,
                 from /usr/local/include/eigen3/Eigen/Core:288,
                 from ~/Open3D/cpp/open3d/geometry/Image.h:29,
                 from ~/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.h:29,
                 from ~/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.cpp:27:
/usr/include/c++/8/bits/stl_algo.h:4330:5: note:   ‘std::transform’
     transform(_InputIterator1 __first1, _InputIterator1 __last1,
     ^~~~~~~~~
In file included from ~/Open3D/cpp/open3d/Open3D.h:40,
                 from ~/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.cpp:29:
~/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
~/Open3D/cpp/open3d/geometry/Line3D.h:374:9: error: ‘transform’ was not declared in this scope
         transform(t);
         ^~~~~~~~~
~/Open3D/cpp/open3d/geometry/Line3D.h:374:9: note: suggested alternative:
In file included from /usr/include/c++/8/algorithm:62,
                 from /usr/local/include/eigen3/Eigen/Core:288,
                 from ~/Open3D/cpp/open3d/geometry/Image.h:29,
                 from ~/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.h:29,
                 from ~/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.cpp:27:
/usr/include/c++/8/bits/stl_algo.h:4330:5: note:   ‘std::transform’
     transform(_InputIterator1 __first1, _InputIterator1 __last1,
     ^~~~~~~~~
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:173: cpp/open3d/geometry/CMakeFiles/geometry.dir/IntersectionTest.cpp.o] Fehler 1
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:368: cpp/open3d/geometry/CMakeFiles/geometry.dir/SurfaceReconstructionBallPivoting.cpp.o] Fehler 1
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:238: cpp/open3d/geometry/CMakeFiles/geometry.dir/MeshBase.cpp.o] Fehler 1
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:498: cpp/open3d/geometry/CMakeFiles/geometry.dir/VoxelGridFactory.cpp.o] Fehler 1
make[2]: *** [cpp/open3d/visualization/CMakeFiles/visualization.dir/build.make:668: cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/RGBDImageShader.cpp.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:1629: cpp/open3d/visualization/CMakeFiles/visualization.dir/all] Fehler 2
make[1]: *** Es wird auf noch nicht beendete Prozesse gewartet....
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:420: cpp/open3d/geometry/CMakeFiles/geometry.dir/TriangleMesh.cpp.o] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:1320: cpp/open3d/geometry/CMakeFiles/geometry.dir/all] Fehler 2
make: *** [Makefile:149: all] Fehler 2

**Expected behavior**
```
`make` should run without this error :-)

**Environment (please complete the following information):**
```
- Operating system:
cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
```
```
$ cat /usr/lib/cuda/version.txt
CUDA Version 10.1.243
```
```
python --version
Python 3.7.3
```
- Open3D version:
0.11.1 
- Remote workstation:no
- How did you install Open3D: build from source
- Compiler version (if built from source): (e.g. gcc 7.5, clang 7.0)
`gcc --version gcc (Debian 8.3.0-6) 8.3.0`
`clang version 7.0.1-8+deb10u2 (tags/RELEASE_701/final)
Target: x86_64-pc-linux-gnu
Thread model: posix`

**Comment by yxlao** [comment]
Looks like it is inherited from `Eigen::ParametrizedLine<double, 3>::transform()`. Added a potential fix in https://github.com/intel-isl/Open3D/pull/2491.

**Comment by nanotuxi** [comment]
Thanks.

**Comment by nanotuxi** [comment]
With your PR #2491 it does not compile. Only changing
`transform(t);` to `Tansform(t);` lets the source compile.
```
In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.cpp:1:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
         this->transform(t);
               ^~~~~~~~~
               Transform
In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/IntersectionTest.h:32,
                 from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/MeshBase.cpp:36:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
         this->transform(t);
               ^~~~~~~~~
               Transform
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:199: cpp/open3d/geometry/CMakeFiles/geometry.dir/Line3D.cpp.o] Fehler 1
make[2]: *** Es wird auf noch nicht beendete Prozesse gewartet....
[ 57%] Building CXX object cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/SimpleBlackShader.cpp.o
In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/IntersectionTest.h:32,
                 from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/IntersectionTest.cpp:27:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
         this->transform(t);
               ^~~~~~~~~
               Transform
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:238: cpp/open3d/geometry/CMakeFiles/geometry.dir/MeshBase.cpp.o] Fehler 1
[ 57%] Building CXX object cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/SimpleShader.cpp.o
make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:173: cpp/open3d/geometry/CMakeFiles/geometry.dir/IntersectionTest.cpp.o] Fehler 1
[ 57%] Building CXX object cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/TexturePhongShader.cpp.o
In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/Open3D.h:40,
                 from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.cpp:29:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
/media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
         this->transform(t);
```

**Comment by manhha1402** [comment]
> With your PR #2491 it does not compile. Only changing
> `transform(t);` to `Transform(t);` lets the source compile.
> 
> ```
> In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.cpp:1:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
>          this->transform(t);
>                ^~~~~~~~~
>                Transform
> In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/IntersectionTest.h:32,
>                  from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/MeshBase.cpp:36:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
>          this->transform(t);
>                ^~~~~~~~~
>                Transform
> make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:199: cpp/open3d/geometry/CMakeFiles/geometry.dir/Line3D.cpp.o] Fehler 1
> make[2]: *** Es wird auf noch nicht beendete Prozesse gewartet....
> [ 57%] Building CXX object cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/SimpleBlackShader.cpp.o
> In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/IntersectionTest.h:32,
>                  from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/IntersectionTest.cpp:27:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
>          this->transform(t);
>                ^~~~~~~~~
>                Transform
> make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:238: cpp/open3d/geometry/CMakeFiles/geometry.dir/MeshBase.cpp.o] Fehler 1
> [ 57%] Building CXX object cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/SimpleShader.cpp.o
> make[2]: *** [cpp/open3d/geometry/CMakeFiles/geometry.dir/build.make:173: cpp/open3d/geometry/CMakeFiles/geometry.dir/IntersectionTest.cpp.o] Fehler 1
> [ 57%] Building CXX object cpp/open3d/visualization/CMakeFiles/visualization.dir/shader/TexturePhongShader.cpp.o
> In file included from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/Open3D.h:40,
>                  from /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/visualization/shader/RGBDImageShader.cpp:29:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h: In member function ‘virtual void open3d::geometry::Segment3D::Transform(const Eigen::Transform<double, 3, 2>&)’:
> /media/tuxi/xcjetson/git_srcs/Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’; did you mean ‘Transform’?
>          this->transform(t);
> ```

I have the same issue with the PR https://github.com/intel-isl/Open3D/pull/2491 building on zotac. Changing `this->transform(t)` to `this->Transform(t)` fix problem

**Comment by nanotuxi** [comment]
@manhha1402 Ooops, yes. Thanks thanks for correcting me ;-) I simply forgot (-> this)

**Comment by Cheng-Hsiung** [comment]
I encounter the same error when building Open3D 0.11.1 with the flag `-DUSE_SYSTEM_EIGEN3=ON`. If I turn this flag off, no error will appear. I thought it was due to eigen version issue so I follow [here ](https://github.com/intel-isl/Open3D/blob/b502d95aabd0f0c486b14935316c205ea494d7e3/3rdparty/README.txt#L12) by pre-installing eigen version 3.3.2 

```
-- Third-Party Dependencies:
--   EIGEN3 .................................. yes (v3.3.2)
...
```
Yet the same error still appears

```
Open3D/cpp/open3d/geometry/Line3D.h:117:15: error: ‘class open3d::geometry::Line3D’ has no member named ‘transform’
         this->transform(t);
Open3D/cpp/open3d/geometry/Line3D.h:374:15: error: ‘class open3d::geometry::Segment3D’ has no member named ‘transform’
         this->transform(t);
```

Is there a fix to this or should I just capitalize the 't' in transform? Many thanks!

**Comment by nanotuxi** [comment]
@Cheng-Hsiung 
As ` yxlao closed this in #2491 12 days ago` you should maybe open a new issue.

The solution to simply `change this->transform(t) to this->Transform(t)` lets the code compile. But there is no guarantee to run into some unknown issues using this  inoffical solution. So better ask @yxlao for an official solution.

**Comment by Cheng-Hsiung** [comment]
Thanks for the reply, @nanotuxi 
Yes at least in my case, I believe this workaround won't compile with `-DUSE_SYSTEM_EIGEN3=OFF`

**Comment by samarth-robo** [comment]
@Cheng-Hsiung @manhha1402 @nanotuxi please see the fix proposed in #2690

**Comment by saedrna** [comment]
I think this problem is because the `transform` function is still not merged in the `3.3` branch of Eigen from its master.
See here [3.3 branch](https://gitlab.com/libeigen/eigen/-/commits/3.3/Eigen/src/Geometry/ParametrizedLine.h)
and here [master](https://gitlab.com/libeigen/eigen/-/commits/master/Eigen/src/Geometry/ParametrizedLine.h)

It is add in commit `d60cca32e54ea02757713935e852cd53343d52c1`.
